### PR TITLE
Add method for library call usage

### PR DESCRIPTION
**Pull request by jpursell**
Add from_points method for NAMatrix struct
Add small example to README.md

### CODE REVIEW

File: src/datastructures/nalgebra.rs [hunk 1]
```diff
@@ -8,6 +9,27 @@ use super::{AdjacencyMatrix, Edge};
 #[derive(Debug, PartialEq)]
 pub struct NAMatrix(pub DMatrix<f64>);
 
+impl NAMatrix {
+    /// Create from set of x/y points
```
Comment (johann-cm) [inline src/datastructures/nalgebra.rs]:
```suggestion
    /// Create from set of x/y points, using the euclidean distance
```

make it more explicit what metric you are using

File: src/datastructures/nalgebra.rs [hunk 2]
```diff
@@ -8,6 +9,27 @@ use super::{AdjacencyMatrix, Edge};
...
+                .par_iter()
+                .skip(i + 1)
+                .map(|to| ((from[0] - to[0]).powi(2) + (from[1] - to[1]).powi(2)).sqrt())
+                .collect();
+            weights.iter().enumerate().for_each(|(j, weight)| {
+                let j = i + j + 1;
+                matrix[(i, j)] = *weight;
+                matrix[(j, i)] = *weight;
+            });
+        }
```
Comment (johann-cm) [inline src/datastructures/nalgebra.rs]:
```suggestion
        points.iter().enumerate().for_each(|(i, from)| {
            points.iter().skip(i + 1).enumerate().for_each(|(j, to)| {
                let j = i + j + 1;
                let weight = ((from[0] - to[0]).powi(2) + (from[1] - to[1]).powi(2)).sqrt();
                // uses the fact, that the euclidean distance is symmetric
                matrix[(i, j)] = weight;
                matrix[(j, i)] = weight;
            });
        });
```
Thanks for your suggestion. I've noticed several things in your code:
1. you are allocating a vector for every iteration of the loop
1. you are using much direct array indexing
1. you are using the parallel iterator only on the inner loop

I suggest the following things:
1. remove the vector allocation, streamline the whole iterator pipeline.
1. transform direct array indexing to the use of iterators
1. remove the parallel iterator, create a separate function for that.

File: src/datastructures/nalgebra.rs [hunk 3]
```diff
@@ -8,6 +9,27 @@ use super::{AdjacencyMatrix, Edge};
...
+        matrix
+    }
```
Comment (johann-cm) [inline src/datastructures/nalgebra.rs:47]:
```suggestion
    }
    /// Create from set of x/y points, using the euclidean distance.
    /// Computation is parallelized via rayon.
    pub fn par_from_points(points: &[[f64; 2]]) -> Self {
        let mut matrix = NAMatrix::from_dim(points.len());
        points
            .par_iter()
            .zip(matrix.par_column_iter_mut())
            .for_each(|(from, mut matrix_col_j)| {
                //let from = &points[i];
                points
                    .iter()
                    .zip(matrix_col_j.iter_mut())
                    .for_each(|(to, matrix_ij)| {
                        let weight = ((from[0] - to[0]).powi(2) + (from[1] - to[1]).powi(2)).sqrt();
                        // cannot use the fact, that the euclidean distance is symmetric,
                        // due to borrowing rules for parallel iterators
                        *matrix_ij = weight;
                    });
            });
        matrix
    }
```



I suggest a separate `par_from_points` function, to let the user opt into rayon parallelism.

One advantage of the parallel implementation is, that it completely removes the use of direct array indexing.
The drawback is, that it does not re-use the computation of `weight` to build the symmetric matrix.

Notice, that the parallel iterator now sits on the outer loop. One could debate, if the inner loop should get a parallel iterator as well.

File: src/datastructures/nalgebra.rs [hunk 4]
```diff
@@ -92,4 +114,18 @@ mod test {
...
+        let matrix = NAMatrix::from_points(&points);
+        for i in 0..points.len() {
+            for j in 0..points.len() {
+                let dist = ((points[i][0] - points[j][0]).powi(2)
+                    + (points[i][1] - points[j][1]).powi(2))
+                .sqrt();
+                assert_eq!(dist, matrix[(i, j)]);
+            }
+        }
+    }
```
Comment (johann-cm) [inline src/datastructures/nalgebra.rs:147]:
```suggestion
    }

    #[test]
    fn test_par_from_points() {
        let points = vec![[0.0, 0.0], [0.0, 1.0], [2.0, 3.0]];
        let matrix = NAMatrix::par_from_points(&points);
        for i in 0..points.len() {
            for j in 0..points.len() {
                let dist = ((points[i][0] - points[j][0]).powi(2)
                    + (points[i][1] - points[j][1]).powi(2))
                .sqrt();
                assert_eq!(dist, matrix[(i, j)]);
            }
        }
    }
```
add test case for the parallel variant

**Review by johann-cm — CHANGES_REQUESTED** [review]
Thank you for your contribution. That is a useful suggestion.


I have some feedback on your implementation and change requests.

**Comment by jpursell** [comment]
That all seems fine to me. I'm curious about what a benchmark would show though, but not so curious I'm going to actually do it. 

I came across this unsafe version. 

https://github.com/1crcbl/cykl-rs/blob/main/src/data.rs#L115

I suspect one could make a parallel version that avoids extra memory allocation and takes advantage of symmetry doing this.

I guess I should also mention that what I'm actually trying to do is TSP art and so far concorde & Lin-Kernighan has been my best option in terms of speed and looks. I just wish I could get this thing to build in windows.

https://github.com/equalis3r/concorde-rs

I'm generating TSPs from around 40,000 points and mostly just don't want the solution to visually cross over itself.

**Review by johann-cm — APPROVED** [review]

**Comment by johann-cm** [comment]
> I'm curious about what a benchmark would show though, but not so curious I'm going to actually do it.

Yep, might be interesting, but I'm also not gonna do it.

Thanks again for your contribution.

**Pull request merged**